src/librustc_session/parse.rs

```diff
@@ -70,6 +72,20 @@ pub enum GateStrength {
     Soft,
```

Comment:
Hmm, looks like we have two mechanisms for soft feature-gating - this one and [the `soft_unstable` lint](https://github.com/rust-lang/rust/issues/64266) - of which this one is unused.

I think we should remove `GateStrength` entirely and simplify the feature gating infra.
"Soft" feature gates are exceptional and having an isolated lint is more than enough.

Reply:
Also, placing the feature gating stuff into `parse.rs` looks strange, it's not related to parsing.

Reply:
> I think we should remove `GateStrength` entirely and simplify the feature gating infra.

Sure, that seems like a good idea, though it seems like an independent change?

> Also, placing the feature gating stuff into `parse.rs` looks strange, it's not related to parsing.

Well... `GatedSpans` is inside `rustc_session::parse`. My goal here is to remove any mentions of `ParseSess` within `syntax` so that I can invert the current dependency on `rustc_session` that `syntax` has and which is bad for parallelism. This `parse.rs` file felt like the best location to put it besides a new module, but I know you don't like those tiny ones. ;)

Reply:
>Sure, that seems like a good idea, though it seems like an independent change?

Yeah, this is material for a different PR (perhaps I'll do it myself today).

>My goal here is to remove any mentions of ParseSess within syntax so that I can invert the current dependency on rustc_session that syntax

I see.

Reply:
So `rustc_session::parse` means `ParseSess` and its methods.
I guess I can live with that, especially given that https://github.com/rust-lang/rust/pull/68018 reduces the surface of feature gating API exposed by `rustc_session::parse`.